pkg/server/settingsworker_test.go

```diff
@@ -293,3 +305,186 @@ func TestSettingsShowAll(t *testing.T) {
+	defer tc.Stopper().Stop(ctx)
+
+	execStatement := func(statement string) {
+		if _, err := tc.ServerConn(rand.Intn(3)).Exec(statement); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Configure tracing with a batch size of 1 in the test cluster
+	execStatement(`SET CLUSTER SETTING trace.zipkin.collector.batch.size = 1`)
+	execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.collector = '%s'`, traceCollectorServerURL.Host))
+
+	// Run statements to be traced
+	execStatement(`CREATE DATABASE t`)
+	execStatement(`CREATE TABLE test (k INT PRIMARY KEY, v INT)`)
+	execStatement(`INSERT INTO test VALUES (1,1)`)
+
+	// Test that the collector observed a batch size of 1 with no error
+	observedBatchReport := <-batchSizeReportingChan
+	if observedBatchReport.Error != nil {
+		t.Fatal(err)
+	}
+	assert.Equal(t, observedBatchReport.ObservedBatchSize, 1)
+}
+
+// TestSettingsTraceSamplingRate tests that setting the sampling rate
+// of traces will correctly limit the percentage of traces sampled.
+func TestSettingsTraceSamplingRate(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// Start test cluster containing a test db and test table
+	ctx := context.Background()
+	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{})
+	defer tc.Stopper().Stop(ctx)
+
+	execStatement := func(statement string) {
+		if _, err := tc.ServerConn(rand.Intn(3)).Exec(statement); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	execStatement(`CREATE DATABASE t`)
+	execStatement(`CREATE TABLE test (k INT PRIMARY KEY, v INT)`)
+
+	// Set up tracing collector and variable to count the number of traces
+	// collected from INSERT INTO statements
+	nTraces := int64(0)
+	collectorHandlerFcn := func(rw http.ResponseWriter, req *http.Request) {
+		bodyBytes, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return
+		}
+		t := thrift.NewTMemoryBuffer()
+		if _, err = t.Write(bodyBytes); err != nil {
+			return
+		}
+		p := thrift.NewTBinaryProtocolTransport(t)
+		_, batchSize, err := p.ReadListBegin()
+		if err != nil {
+			return
+		}
+
+		for i := 0; i < batchSize; i += 1 {
+			span := zipkincore.Span{}
+			if err = span.Read(p); err != nil {
+				return
+			}
+
+			if span.Name == `exec stmt` {
+				for _, annotation := range span.Annotations {
+					if strings.Contains(annotation.Value, `INSERT INTO test VALUES`) &&
+						strings.Contains(annotation.Value, `in state: Open`) {
+						atomic.AddInt64(&nTraces, 1)
+					}
+				}
+			}
+		}
+	}
+	traceCollectorServer := httptest.NewServer(http.HandlerFunc(collectorHandlerFcn))
+	defer traceCollectorServer.Close()
+
+	// Configure tracing in the test cluster
+	traceCollectorServerURL, err := url.Parse(traceCollectorServer.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+	setClusterTraceSamplingRate := func(rate float64) {
+		execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.sampling.rate = %f`, rate))
+	}
+	execStatement(`SET CLUSTER SETTING trace.zipkin.collector.batch.size = 100`)
```

Comment:
I actually did not end up setting the batch size to 0 to make the traces sent to the collector more synchronous. When writing 1000 rows in parallel, the collector simply drops traces if the buffer (which uses the same setting as the batch size) is full. Thus, the batch size needs to be large so traces don't get dropped and we can count them.